File: clj-lein/src/co/paralleluniverse/quasar_app.clj
(ns co.paralleluniverse.quasar-app
	(:gen-class)
	(:require
		[co.paralleluniverse.pulsar.core :refer [channel fiber snd rcv join close!]])
	(:import (co.paralleluniverse.fibers Fiber)))

(defn -main
	"Application entry point"
	[& args]
	(let
		[increasing-to-echo (channel)
		 echo-to-increasing (channel)
		 increasing-fiber (fiber
												(loop [i 0
															 curr 0]
													(if (< i 10)
														(do
															(Fiber/sleep 1000)
															(println "INCREASER sending:" curr)
															(snd increasing-to-echo curr)
															(let [curr (rcv echo-to-increasing)]
																(println "INCREASER received:" curr)
																(println "INCREASER now:" (+ curr 1))
																(recur (+ i 1)
																			 (+ curr 1))))
														(do
															(println "INCREASER closing channel and exiting")
															(close! increasing-to-echo)))))
		 echo-fiber (fiber
									(loop []
										(do
											(Fiber/sleep 1000)
											(let [curr (rcv increasing-to-echo)]
												(println "ECHO received:" curr)
												(if curr
													(do
														(println "ECHO sending:" curr)
														(snd echo-to-increasing curr)
														(recur))
													(do
														(println "ECHO detected closed channel, closing and exiting")
														(close! echo-to-increasing)))))))]
		(join increasing-fiber)
		(join echo-fiber)))

File: src/main/java/co/paralleluniverse/QuasarApp.java
package co.paralleluniverse;

import java.util.concurrent.ExecutionException;

import co.paralleluniverse.strands.channels.Channels;
import co.paralleluniverse.strands.channels.IntChannel;

import co.paralleluniverse.fibers.Fiber;

/**
 * Increasing-Echo Quasar Example
 *
 * @author circlespainter
 */
public class QuasarApp {
    public static void main(String[] args) throws ExecutionException, InterruptedException {
        final IntChannel increasingToEcho = Channels.newIntChannel(0); // Synchronizing channel (buffer = 0)
        final IntChannel echoToIncreasing = Channels.newIntChannel(0); // Synchronizing channel (buffer = 0)

        final Fiber<Integer> increasing = new Fiber<>("INCREASER", () -> {
            ////// The following is enough to test instrumentation of synchronizing methods
            // synchronized(new Object()) {}

            int curr = 0;
            for (int i = 0; i < 10 ; i++) {
                Fiber.sleep(1000);
                System.out.println("INCREASER sending: " + curr);
                increasingToEcho.send(curr);
                curr = echoToIncreasing.receive();
                System.out.println("INCREASER received: " + curr);
                curr++;
                System.out.println("INCREASER now: " + curr);
            }
            System.out.println("INCREASER closing channel and exiting");
            increasingToEcho.close();
            return curr;
        }).start();

        final Fiber<Void> echo = new Fiber<Void>("ECHO", () -> {
            Integer curr;
            while (true) {
                Fiber.sleep(1000);
                curr = increasingToEcho.receive();
                System.out.println("ECHO received: " + curr);

                if (curr != null) {
                    System.out.println("ECHO sending: " + curr);
                    echoToIncreasing.send(curr);
                } else {
                    System.out.println("ECHO detected closed channel, closing and exiting");
                    echoToIncreasing.close();
                    return;
                }
            }
        }).start();

        try {
            increasing.join();
            echo.join();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
    }
}
